res/controllers/lib/mididispatcher.mjs

```diff
@@ -0,0 +1,89 @@
+/**
+* MidiDispatcher routes incoming MIDI messages to registered callbacks. Register callbacks with
+* the setInputCallback method in your controller script module's init function and call
+* MidiDispatcher's receiveMidiData method in your controller script module's receiveData function.
+*
+* This class is not designed to handle System Exclusive or MIDI Clock messages. For those, implement logic
+* specific to your controller in the controller module's receiveData function before calling
+* MidiDispatcher.receiveData.
+*/
+
+// MIDI messages starting with 0xC (program change) or 0xD (aftertouch) messages are only
+// two bytes long and distinguished by their first byte.
+// https://www.midi.org/specifications-old/item/table-2-expanded-messages-list-status-bytes
+const identifiedByFirstByte = (midiBytes) => {
+    return (midiBytes[0] & 0xF0) == 0xC0 || (midiBytes[0] & 0xF0) === 0xD0;
+}
+
+// JavaScript is broken and believes [1,2] === [1,2] is false, so
+// this function turns an array of MIDI bytes into a unique hashable key.
+// This function does not work for system exclusive messages; that is out of scope.
+const hashMidiBytes = (midiBytes) => {
+    if (identifiedByFirstByte(midiBytes)) {
+        return midiBytes[0];
+    }
+    return midiBytes[0] + (midiBytes[1] << 8);
+}
+
+export class MidiDispatcher {
+    /**
+     * @param {bool} noteOff - When setting the callback for a Note On message, also map the corresponding Note Off
+     * message to the same callback. Defaults to true.
+     */
+    constructor(noteOff) {
+        if (noteOff === undefined) {
+            noteOff = true;
+        }
+        this.noteOff = noteOff;
+        this.inputMap = new Map();
+    }
+    /**
+     * Set the callback for an incoming MIDI message.
+     * @param {Array} midiBytes - Array of numbers indicating the beginning of the MIDI messages. In most cases,
```

Comment:
```suggestion
     * @param {number[]} midiBytes - Array of numbers indicating the beginning of the MIDI messages. In most cases,
```
just says its an array (making its elements implicitly `any`). Since we're doing some bitshifting, we should specify that we're expecting numbers.

Reply:
Also, same goes here in regards to using `Uint8Array` here